bigquery_etl/publish_json.py

```diff
@@ -133,26 +133,57 @@ def _gcp_convert_ndjson_to_json(self, gcs_path):
         )
         bucket = self.storage_client.bucket(self.target_bucket)
 
+        # keeps track of the number of bytes written to the JSON output file
+        output_size = 0
+        # tracks the current output file name
+        output_file_counter = 0
+        # track if the first JSON object is currently processed
+        first_line = True
+        # output file handler
+        output_file = None
+
         for blob in blobs:
             blob_path = f"gs://{self.target_bucket}/{blob.name}"
-            tmp_blob_name = blob_path + ".tmp.gz"
-            tmp_blob_name = tmp_blob_name.replace("ndjson", "json")
-
-            logging.info(f"""Compress {blob_path} to {tmp_blob_name}""")
 
             # stream from GCS
             with smart_open.open(blob_path) as fin:
-                with smart_open.open(tmp_blob_name, "w") as fout:
-                    fout.write("[\n")
-
-                    for i, line in enumerate(fin):
-                        # skip the first line, it has no preceding json object
-                        if i > 0:
-                            fout.write(",\n")
-
-                        fout.write(line.replace("\n", ""))
-
-                    fout.write("]")
+                for line in fin:
+                    # check if JSON output file reached limit
+                    if output_file is None or output_size >= MAX_JSON_SIZE:
+                        if output_file is not None:
+                            output_file.write("]")
+                            output_file.close()
+
+                        if output_file_counter >= 1000000000000:
+                            logging.error(
+                                "Maximum number of JSON output files reached."
+                            )
+                            sys.exit(1)
+
+                        tmp_blob_name = "/".join(blob.name.split("/")[:-1])
+                        tmp_blob_name += (
+                            "/" + str(output_file_counter).zfill(12) + ".json.tmp.gz"
+                        )
+                        tmp_blob_path = f"gs://{self.target_bucket}/{tmp_blob_name}"
+
+                        logging.info(f"""Write {blob_path} to {tmp_blob_path}""")
+
+                        output_file = smart_open.open(tmp_blob_path, "w")
+                        output_file.write("[\n")
```

Comment:
We could alleviate some size constraints by eliding the newlines. JSON viewers in browsers should be an acceptable alternative.